qudipy/qutils/matrices.py

```diff
@@ -1,315 +1,416 @@
-    """
-    Defines a raising operator of the k-th qubit
-
-    Parameters
-    ----------
-    N : int
-        Number of 1-qubit degrees of freedom in the operator.
-    k : int
-        Position of the raising operator in the tensor product.
-
-    Returns
-    -------
-    : complex 2D array
-        The raising operator X_k + i Y_k
-    """
-    return x(N, k) + complex(0.0, 1.0) * y(N, k)
-
-
-def sigma_minus(N, k):
-    """
-    Defines a lowering operator of the k-th qubit
-
-    Parameters
-    ----------
-    N : int
-        Number of 1-qubit degrees of freedom in the operator.
-    k : int
-        Position of the raising operator in the tensor product.
-
-    Returns
-    -------
-    : complex 2D array
-        The lowering operator X_k - i Y_k
-    """
-    return x(N, k) - complex(0.0, 1.0) * y(N, k)
-    
-def e_up(N, k):
-    """
-    Defines matrix that projects k-th qubit on the state |0〉
-    
-    Parameters
-    ----------
-    N : int
-        Number of 1-qubit degrees of freedom in the operator.
-    k : int
-        Position of the projection matrix in the tensor product.
+            return u_k
+
+    ##### Hard Coded Operators #####
+
+    # Ladder operatorss X_k ± i Y_k
+    def sigma_plus(self,N, k):
+        """
+        Defines a raising operators of the k-th qubit
+        Parameters
+        ----------
+        N : int
+            Number of 1-qubit degrees of freedom in the operators.
+        k : int
+            Position of the raising operators in the tensor product.
+        Returns
+        -------
+        : complex 2D array
+            The raising operators X_k + i Y_k
+        """
+        return (self.construct(N,k,self.operators['PAULI_X']) 
+            + complex(0.0, 1.0) * self.construct(N,k,self.operators['PAULI_Y']))
+
+    def sigma_minus(self,N, k):
+        """
+        Defines a lowering operators of the k-th qubit
+        Parameters
+        ----------
+        N : int
+            Number of 1-qubit degrees of freedom in the operators.
+        k : int
+            Position of the raising operators in the tensor product.
+        Returns
+        -------
+        : complex 2D array
+            The lowering operators X_k - i Y_k
+        """
+        return (self.construct(N,k,self.operators['PAULI_X']) 
+            - complex(0.0, 1.0) * self.construct(N,k,self.operators['PAULI_Y']))
         
-    Keyword Arguments
-    -----------------
-    None.
-    
-    Returns
-    -------
-    : 2D complex array
-        Matrix |0〉〈0|_k of dimensions 2**N x 2**N 
-
-    """
-    return 0.5 * (unit(N) + z(N, k))
-
-
-def e_down(N, k):
-    """
-    Defines matrix that projects k-th qubit on the state |1〉
-    
-    Parameters
-    ----------
-    N : int
-        Number of 1-qubit degrees of freedom in the operator.
-    k : int
-        Position of the projection matrix in the tensor product.
+    def e_up(self,N, k):
+        """
+        Defines matrix that projects k-th qubit on the state |0〉
```

Comment:
I used this elsewhere in the code but since the name is "e_up", the ket state should look like |↑〉.
To fix it only in one place, I suggest the following: 
```suggestion
        Defines matrix that projects k-th qubit on the state |↑〉≡ |0〉
```